feature(hx): add cross-domain scenario.

XinHao-96 opened the pull request:
A pipeline for federated learning in cross-domain scenarios. In this setup, client data originates from different domains (it should be noted that in the cross-domain pipeline, the hyperparameter --client_num refers to the number of clients into which data from the same domain has been divided). Final training performance is evaluated locally on each client (it is important to note that clients within the same domain share the same test data).

Dataset: A new function, `get_cross_domain_dataset`, has been created to load datasets based on domain information. Additionally, a new cross-domain dataset, DomainNet, has been included.

Data Sampling: A conditional check for cross-domain datasets has been added, ensuring that clients from the same domain share the same test set.

Group: A new class, `CrossDomainServerGroup`, has been created, inheriting from the `ParameterServerGroup` parent class. The primary modification involves indexing clients based on domain information and the number of clients per domain. To support this, a new `cross-domain-avg` method was also implemented to enable proper aggregation for clients with domain-specific information.

Server: A new class, `CrossDomainServer`, was created, inheriting from `ServerTemplate`. The primary modification involves adding indexing based on domain information (the server testing method from `generic_model_pipeline` can be used here).

Client: To enable precise localization of cross-domain clients, a combined indexing method using both domain name and client ID has been implemented. A new function, `get_cross_domain_client`, was created to obtain these two parameters. Additionally, a `CrossDomainClient` class, designed for cross-domain scenarios, was introduced, inheriting from `ClientTemplate`.
                                                                                                                                                                                          --Xin Hao

XinHao-96 committed:
Cross-Domain

kxzxvbk commented on fling/component/client/build_client.py:
```diff
@@ -15,3 +15,15 @@ def get_client(
         test_dataset=test_dataset,
         **kwargs
     )
+def get_cross_domain_client(
```
Maybe there is no need to add this function. You can use ``**kwargs`` to pass the argument domain.

kxzxvbk commented on fling/component/client/cross_domain_client.py:
```diff
@@ -0,0 +1,235 @@
+
+
+@CLIENT_REGISTRY.register('cross_domain_client')
+class CrossDomainClient(ClientTemplate):
```
Maybe this class can inherit BaseClient. I think there is no difference in function train, test, finetune, preprocess_data. And the __init__ function can be largely simplified by calling super.__init__()

kxzxvbk commented on fling/component/group/cross_domain_group.py:
```diff
@@ -0,0 +1,149 @@
+            'Parameter number in each model: {:.2f}M'.format(get_params_number(self.clients[self.domains[0]][0].model) / 1e6)
+        )
+    
+    def append(self, domain: str, client: ClientTemplate) -> None:
```
Remove this function?

kxzxvbk commented on fling/component/group/cross_domain_group.py:
```diff
@@ -0,0 +1,149 @@
+
+        return trans_cost
+
+    def flush(self) -> None:
```
Remove this function?

kxzxvbk commented on fling/component/server/cross_domain_server.py:
```diff
@@ -0,0 +1,63 @@
+            weight=batch_y.shape[0]
+        )
+    
+    def preprocess_data(self, data):
```
Maybe remove this function.

kxzxvbk commented on flzoo/domainnet/domainnet_fedavg_resnet_config.py:
```diff
@@ -0,0 +1,37 @@
+        input_channel=3,
+        class_number=10,
+    ),
+    client=dict(name='cross_domain_client', client_num=2),
```
Give more detailed comments about the meaning of `client_num` here, as it is not the same meaning as that in typical FL.

kxzxvbk commented:
Reformat your code by:
```
source ./format.sh ./
flake8 ./
```

kxzxvbk commented:
Add cross-domain information in the README file.

XinHao-96 committed:
New Edition

XinHao-96 committed:
Delete flzoo/domainnet/domainnet_fedavg_resnet_config_local.py

XinHao-96 commented:
Based on your suggestions, I have made the following modifications to the original code:

Changed `cross_domain_client` to inherit from `BaseClient` and removed redundant parts.
Removed redundant sections in `cross_domain_group`.
Changed `cross_domain_server` to inherit from `BaseServer` and eliminated duplicate code.
Reformatted the parts related to cross-domain modifications using `source ./format.sh ./`. During the actual run, I also made formatting adjustments to other parts, such as the `FedPart` code, but to avoid confusion, I have not reflected these changes in the pull request.

One part that still confuses me is that the Fling README file does not seem to contain a description of heterogeneity or the scenarios. Therefore, I am unsure where it would be most appropriate to place the introduction to cross-domain.

kxzxvbk commented on fling/dataset/build_dataset.py:
```diff
@@ -5,3 +5,7 @@
     return DATASET_REGISTRY.build(args.data.dataset, cfg=args, train=train, **kwargs)
+
+
+def get_cross_domain_dataset(args: dict, domain: str, train: bool, **kwargs) -> Dataset:
```
Maybe this function can be directly removed, and simply use ``get_dataset``?

XinHao-96 committed:
get dataset

XinHao-96 commented:
removed the get_cross_domain_dataset function by using get_dataset function instead.
i'm still working on enlarging readme file.

XinHao-96 commented:
removed the get_cross_domain_dataset function by using get_dataset function instead.
i'm still working on enlarging readme file.

XinHao-96 committed:
Update README.md

XinHao-96 committed:
readme and demo

XinHao-96 committed:
Update README.md

XinHao-96 committed:
Merge branch 'main' into main

XinHao-96 commented:
I enlarged the cross-domain description part in readme.md and created two demos by using FedPer and FedBN algorithm with DomainNet dataset.

XinHao-96 committed:
data loading

XinHao-96 committed:
domainnet dataset preparation

XinHao-96 commented:
I followed the instructions in the file to download DomainNet, and then I ran the test locally. It worked fine.
